staging/src/k8s.io/kube-apiextensions-server/pkg/apis/apiextensions/types.go

```diff
@@ -0,0 +1,129 @@
+	// Names are the names used to describe this custom resource
+	Names CustomResourceNames
+
+	// Scope indicates whether this resource is cluster or namespace scoped.  Default is namespaced
+	Scope ResourceScope
+}
+
+// CustomResourceNames indicates the names to serve this CustomResource
+type CustomResourceNames struct {
+	// Plural is the plural name of the resource to serve.  It must match the name of the TPR-registration
+	// too: plural.group
+	Plural string
+	// Singular is the singular name of the resource.  Defaults to lowercased <kind>
+	Singular string
+	// ShortNames are short names for the resource.
+	ShortNames []string
+	// Kind is the serialized kind of the resource
+	Kind string
+	// ListKind is the serialized kind of the list for this resource.  Defaults to <kind>List
+	ListKind string
+}
+
+// ResourceScope is an enum defining the different scopes availabe to a custom resource
+type ResourceScope string
+
+const (
+	ClusterScoped   ResourceScope = "Cluster"
+	NamespaceScoped ResourceScope = "Namespaced"
+)
+
+type ConditionStatus string
+
+// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
+// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
+// can't decide if a resource is in the condition or not. In the future, we could add other
+// intermediate conditions, e.g. ConditionDegraded.
+const (
+	ConditionTrue    ConditionStatus = "True"
+	ConditionFalse   ConditionStatus = "False"
+	ConditionUnknown ConditionStatus = "Unknown"
+)
+
+// CustomResourceConditionType is a valid value for CustomResourceCondition.Type
+type CustomResourceConditionType string
+
+const (
+	// NameConflict means the names chosen for this CustomResource conflict with others in the group.
+	NameConflict CustomResourceConditionType = "NameConflict"
+	// Terminating means that the CustomResource has been deleted and is cleaning up.
+	Terminating CustomResourceConditionType = "Terminating"
+)
+
+// CustomResourceCondition contains details for the current condition of this pod.
+type CustomResourceCondition struct {
+	// Type is the type of the condition.
+	Type CustomResourceConditionType
+	// Status is the status of the condition.
+	// Can be True, False, Unknown.
+	Status ConditionStatus
+	// Last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// Unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string
+	// Human-readable message indicating details about last transition.
+	// +optional
+	Message string
+}
+
+// CustomResourceStatus indicates the state of the CustomResource
+type CustomResourceStatus struct {
+	// Conditions indicate state for particular aspects of a CustomResource
+	Conditions []CustomResourceCondition
+
+	// AcceptedNames are the names that are actually being used to serve discovery
+	// They may be different than the names in spec.
```

Comment:
Forseeing clients which do complicated comparison logic whether everything specified is actually accepted. Can we at least express this with a boolean `Accepted: Partially/Complete` ?